src/brod_client.erl

```diff
@@ -418,21 +419,23 @@ shutdown_pid(_) ->
 
 -spec get_partition_worker(client(), partition_worker_key()) ->
         {ok, pid()} | {error, get_worker_error()}.
-get_partition_worker(ClientPid, Key) when is_pid(ClientPid) ->
-  case erlang:process_info(ClientPid, registered_name) of
-    {registered_name, ClientId} when is_atom(ClientId) ->
-      get_partition_worker(ClientId, Key);
-    _ ->
-      %% This is a client process started without registered name
-      %% have to call the process to get the producer/consumer worker
-      %% process registration table.
-      case safe_gen_call(ClientPid, get_workers_table, infinity) of
-        {ok, Ets}       -> lookup_partition_worker(ClientPid, Ets, Key);
-        {error, Reason} -> {error, Reason}
-      end
-  end;
-get_partition_worker(ClientId, Key) when is_atom(ClientId) ->
-  lookup_partition_worker(ClientId, ?ETS(ClientId), Key).
+get_partition_worker(Client, Key) ->
+  %% Added potentially gratuitous call to `get_workers_table` to flush Client inbox
+  %% which ensures that ets is in sync with the children being supervised.
+  case {is_atom(Client), safe_gen_call(Client, get_workers_table, infinity)}  of
```

Comment:
This is a very hot code path for producers,
the unconditional gen call is likely a performance killer.

Maybe we can make gen call if:
Lookup failed, or lookup returned a process which is `not erlang:is_process_alive(Pid)`
So the overhead added is, in most of the cases, just `is_process_alive`

A side note: all “worker discovery” have to face the race condition of: work process being down after discovery when submitting a task to it.
This change only lowers the chance of such race.

Reply:
100%, My understanding is that just because we "flush" the inbox right before we call ETS doesn't mean that something else couldn't sneak in right before we actually look at ETS.

Reply:
Curious if you have any insight into the note about [supervisor:which_children/1](https://www.erlang.org/doc/man/supervisor.html#which_children-1)
>Notice that calling this function when supervising many children under low memory conditions can cause an out of memory exception.

It seems like being able to use that function would solve so many headaches in so many scenarios. But because of the vagueness around _many_ and _low_ in the above statement it appears to be off limits.